### PR TITLE
mentions in groups and forums to only members of that group

### PR DESCRIPTION
**Pull request by mehulkaklotar**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Mention only group members in group and forums.
https://files.slack.com/files-pri/T0VNCRXSL-F011LFTPW3G/image_from_ios.jpg

### How to test the changes in this Pull Request:

1. Go to group > feed
2. try to mention in activity feed
3. It should only allow members of that group.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Mentions - groups and forums mentions results restricted to members of group.

### CODE REVIEW

**Comment by eisenwasser** [comment]
@ankitguptaindia 

Can you deep test this, and let me know when its ready for me to test. Thanks.

File: src/bp-groups/bp-groups-functions.php [hunk 1]
```diff
@@ -3766,3 +3766,55 @@ function groups_can_user_manage_messages( $user_id, $group_id ) {
 
 	return $is_allowed;
 }
+
+function bp_groups_prime_mentions_results() {
```
Comment (chetansatasiya) [inline src/bp-groups/bp-groups-functions.php:3770]: @mehulkaklotar It would be great we can add doc block.

**Comment by chetansatasiya** [comment]
Issue video: https://www.loom.com/share/b84414e86b464cd3b4eea47a3cc4f278

File: src/bp-groups/bp-groups-functions.php (hunk 1 above)
Comment (mehulkaklotar) [reply src/bp-groups/bp-groups-functions.php:3775]: done.

**Comment by mehulkaklotar** [comment]
> @mehulkaklotar Code looks good to me but the issue is still there like If I tag other people into the Group forum discussion OR in reply it will allow me to tag the people which is not a part of that group.

That is not the issue. here we are just filtering the mnetions results when we search for them. In site, if you add `@mention_user`. WordPress converts them automatically for posts.

**Comment by ankitguptaindia** [comment]
@mehulkaklotar @chetansatasiya 

Issue 1. - Test Scenario- -  when group has many user and logged-in User A tagged to User B and on the other hand If user B is logged-in and want to Tag User A. It should not show self user as mention suggestion AND ISSUE is- it is not showing other member in suggestion.  

https://prnt.sc/s33sj1
https://www.loom.com/share/632bb474f662480aba850639c71e0c89
https://prnt.sc/s33ol3

Issue -2. In Private Group Forum Discussion, it shows all site members Although Group has not any member in Group.
https://www.loom.com/share/ab85bbf167f94f60bcf86ac54fe17c6f


Tested scenario/use cases in :-

- tested in Public Group
- tested in Private Group
- tested in Hidden Group
- Tested in Group which has Parent-Child hierarchy 
- Tested when Activity Component is OFF.

**Comment by mehulkaklotar** [comment]
> @mehulkaklotar @chetansatasiya
> 
> Issue 1. - Test Scenario- - when group has many user and logged-in User A tagged to User B and on the other hand If user B is logged-in and want to Tag User A. It should not show self user as mention suggestion AND ISSUE is- it is not showing other member in suggestion.
> 
> https://prnt.sc/s33sj1
> https://www.loom.com/share/632bb474f662480aba850639c71e0c89
> https://prnt.sc/s33ol3
> 
> Issue -2. In Private Group Forum Discussion, it shows all site members Although Group has not any member in Group.
> https://www.loom.com/share/ab85bbf167f94f60bcf86ac54fe17c6f
> 
> Tested scenario/use cases in :-
> 
> * tested in Public Group
> * tested in Private Group
> * tested in Hidden Group
> * Tested in Group which has Parent-Child hierarchy
> * Tested when Activity Component is OFF.

Case 1: I don't know what the issue here. It should show group members only and it is showing.
Case 2: Fixed

**Comment by ankitguptaindia** [comment]
@mehulkaklotar Urgent Please...

When user getting notification about mentioning in Private Group then that notification shows no activity found on click. Albeit user is member of group.

https://www.loom.com/share/79566d2fa798420ba6f21a927e326379

**Comment by ankitguptaindia** [comment]
@mehulkaklotar 

User can mention any NON MEMBER in Group from news feed page..
Is it expected from news feed or we can make condition for this scenario?

https://www.loom.com/share/ce2e9e21d7174caf9f95dd2d113ffb15

**Comment by mehulkaklotar** [comment]
> @mehulkaklotar Urgent Please...
> 
> When user getting notification about mentioning in Private Group then that notification shows no activity found on click. Albeit user is member of group.
> 
> https://www.loom.com/share/79566d2fa798420ba6f21a927e326379

@ankitguptaindia this is from activity privacy.


> @mehulkaklotar
> 
> User can mention any NON MEMBER in Group from news feed page..
> Is it expected from news feed or we can make condition for this scenario?
> 
> https://www.loom.com/share/ce2e9e21d7174caf9f95dd2d113ffb15

@ankitguptaindia on news feed page, we can't block users to mention other than groups because we dont have group members on the page yet. user can mention friends or members on the post content on news feed.

**Comment by ankitguptaindia** [comment]
@mehulkaklotar 
Self Notification getting trigger when user tag himself.  **(If this take time we can fix this in later in new PR)**
https://www.loom.com/share/a1d8ba185f814b8d926752d03d0e6ed4

**Comment by ankitguptaindia** [comment]
@mehulkaklotar 

Group Member can mention forcefully to any NON GROUP MEMBER by typing some random username

for ex. john is not member of group but I can type @john and it converts this to mentions.

https://www.loom.com/share/7f5c847589bf4b368e19059cd14f9c3b

**Comment by mehulkaklotar** [comment]
> @mehulkaklotar
> Self Notification getting trigger when user tag himself. **(If this take time we can fix this in later in new PR)**
> https://www.loom.com/share/a1d8ba185f814b8d926752d03d0e6ed4

@ankitguptaindia If user tags himself forcefully then this is not our concern. Notification should generate whenever user is mentioned in the activity because he has enabled the settings for it.


> @mehulkaklotar
> 
> Group Member can mention forcefully to any NON GROUP MEMBER by typing some random username
> 
> for ex. john is not member of group but I can type @john and it converts this to mentions.
> 
> https://www.loom.com/share/7f5c847589bf4b368e19059cd14f9c3b

@ankitguptaindia  Non member of public group can be mentioned in the news feed or in the group feed even if it is not in the suggestion list. Private and Hidden group will not generate any mention notification for the non member mention.

**Comment by eisenwasser** [comment]
@mehulkaklotar @ankitguptaindia 

What is the status on this one?

**Comment by ankitguptaindia** [comment]
> @mehulkaklotar @ankitguptaindia
> 
> What is the status on this one?

@eisenwasser I will review this after LifterLms tasks.
This is next in my list for review.

**Comment by eisenwasser** [comment]
Ok thanks

**Comment by ankitguptaindia** [comment]
@mehulkaklotar 
Thanks for the clarification of these scenario.

- https://github.com/buddyboss/buddyboss-platform/pull/747#issuecomment-645168265
- https://github.com/buddyboss/buddyboss-platform/pull/747#issuecomment-644559316

Others issues have been fixed. 👍

**Pull request merged**